### PR TITLE
rgw/s3select: json integration

### PR DESCRIPTION
**Pull request by albin-antony**
Integration of json processing by s3select engine

Signed-off-by: Albin Antony <aantony@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by albin-antony** [comment]
Thank you @galsalomon66. Please take a look

File: src/rgw/rgw_s3select.cc [hunk 1]
```diff
@@ -460,6 +463,58 @@ int RGWSelectObj_ObjStore_S3::run_s3select_on_parquet(const char* query)
...
+    m_aws_response_handler.init_response();
+    if (input == nullptr) {
+      input = "";
+    }
+    std::string temp_result;
+    m_aws_response_handler.init_success_response();
+    length_before_processing = temp_result.size();
+    //query is correct(syntax), processing is starting.
+    status = m_s3_json_object.run_s3select_on_stream(temp_result, input, input_length, s->obj_size);
+    m_aws_response_handler.get_sql_result().append(temp_result);
```
Comment (galsalomon66) [inline src/rgw/rgw_s3select.cc:483]: ok.
`temp_result` implies it's a temporary resolution.
upon this issue (override the SQL result) is resolved, this code should be re-factored.

File: src/rgw/rgw_s3select.cc [hunk 2]
```diff
@@ -485,6 +540,10 @@ int RGWSelectObj_ObjStore_S3::handle_aws_cli_parameters(std::string& sql_query)
   extract_by_tag(m_s3select_input, "QuoteCharacter", m_quot);
   extract_by_tag(m_s3select_input, "RecordDelimiter", m_row_delimiter);
   extract_by_tag(m_s3select_input, "FileHeaderInfo", m_header_info);
+  extract_by_tag(m_s3select_input, "Type", m_json_datatype);
```
Comment (galsalomon66) [inline src/rgw/rgw_s3select.cc:543]: attached is the message delivered by the client-side (`AWS-CLI` --> `RGW`)

Content-Length: 363\r\n\r\n\<SelectObjectContentRequest xmlns="[http://s3.amazonaws.com/doc/2006-03-01/](http://s3.amazonaws.com/doc/2006-03-01/%5C)"\>\<Expression\>select count(0) from s3object[*];\</Expression\>\<ExpressionType\>SQL\</ExpressionType\>\<InputSerialization\>\<JSON\>\<Type\>DOCUMENT\</Type\>\</JSON\>\<CompressionType\>NONE\</CompressionType\>\</InputSerialization\>\<OutputSerialization\>\<CSV /\>\</OutputSerialization\>\</SelectObjectContentRequest\>"

the **InputSerialization** tag contains the object type (CSV / JSON / Parquet), the flow should use that.
the **Type** tag is a sub-tag (in the case of JSON it's "Document" or "Lines")

**Comment by galsalomon66** [comment]
if there is a modification for s3selelect-repo, it should be noted in this PR.

the PR failed the build https://jenkins.ceph.com/job/ceph-pull-requests/107855/consoleFull#14557452702a811ea2-3e7b-466b-84b4-d13df7e35809


this PR should be tested for JSON integration (RGW can process a JSON object correctly)
few tests are enough. (the test PR should be noted in this PR)

**Comment by albin-antony** [comment]
> > This pull request can no longer be automatically merged: a rebase is needed and changes have to be manually resolved
> 
> @albin-antony the commit for s3select submodule is wrong it contains nonrelevant changes (probably the s3select PR was not updated, the PR should contain only the relevant changes)

I am on it

**Comment by galsalomon66** [comment]
@albin-antony 
please check conflict

**Comment by albin-antony** [comment]
#49411

**Pull request closed without merging**